airbyte-integrations/connectors/source-flexport/source_flexport/source.py

```diff
@@ -0,0 +1,198 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple, Union
+from urllib.parse import parse_qsl, urlparse
+
+import pendulum
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth.core import HttpAuthenticator
+from airbyte_cdk.sources.streams.http.requests_native_auth import TokenAuthenticator
+from requests.auth import AuthBase
+
+
+class FlexportError(Exception):
+    pass
+
+
+class FlexportStream(HttpStream, ABC):
+    url_base = "https://api.flexport.com/"
+    raise_on_http_errors = False
+    primary_key = "id"
+    page_size = 500
+
+    def __init__(self, authenticator: Union[AuthBase, HttpAuthenticator] = None, start_date: str = None):
+        super().__init__(authenticator=authenticator)
+
+        self._authenticator = authenticator
+        self.start_date = start_date
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        # https://apidocs.flexport.com/reference/pagination
+        # All list endpoints return paginated responses. The response object contains
+        # elements of the current page, and links to the previous and next pages.
+        data = response.json()["data"]
+
+        if data["next"]:
+            url = urlparse(data["next"])
+            qs = dict(parse_qsl(url.query))
+
+            return {
+                "page": qs["page"],
+                "per": qs["per"],
+            }
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        if next_page_token:
+            return next_page_token
+
+        return {
+            "page": 1,
+            "per": self.page_size,
+        }
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        # https://apidocs.flexport.com/reference/response-layout
+        json = response.json()
+
+        http_error = None
+        try:
+            response.raise_for_status()
+        except Exception as exc:
+            http_error = exc
+
+        error = json.get("error")
+        if error:
+            if "code" in error and "message" in error:
+                raise FlexportError(f"{error['code']}: {error['message']}") from http_error
+
+            if not http_error:
+                raise FlexportError(f"Unexpected error: {error}")
```

Comment:
I'd rather raise the original request exception, wdyt?

Reply:
I updated error handling to be more defensive. The algorithm is following:

1. Check for the `error` property.
1.1. Raise if the object is not a dict.
2. Check for the contents of the `error` property.
2.1 Try to raise the correct error.	
2.2. Try to infer the error message from the malformed error object.
3. Raise plain HTTP error, if any.

In addition, I updated test case ordering to match the algorithm, see `test_parse_response`.